IntegrationTests/UI.IntegrationTests/UITest.cs

```diff
@@ -27,32 +28,69 @@ void HandleApplicationIdle(object sender, EventArgs e)
+            try
             {
-                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
-                await WaitForIdleAsync();
-                var dialog = Application.OpenForms.OfType<T>().Single();
-                try
-                {
-                    await runAsync(dialog);
-                }
-                finally
+                // Start runTestAsync before calling showForm.
+                // The latter might block until the form is closed, especially if using Application.Run(form).
+
+                // Avoid using ThreadHelper.JoinableTaskFactory for the outermost operation because we don't want the task
+                // tracked by its collection. Otherwise, test code would not be able to wait for pending operations to
+                // complete.
+                var test = ThreadHelper.JoinableTaskContext.Factory.RunAsync(async () =>
                 {
-                    dialog.Close();
-                }
-            });
+                    try
+                    {
+                        await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                        await WaitForIdleAsync();
+                        form = Application.OpenForms.OfType<T>().Single();
+                        await runTestAsync(form);
+                    }
+                    finally
+                    {
+                        try
+                        {
+                            form?.Close();
+                        }
+                        finally
+                        {
+                            // Try to close all open Forms in order to let return the blocking Application.Run(form)
+                            // which might have been called by showForm() below.
+                            for (int trial = 0; trial < 3 && Application.OpenForms.Count > 0; ++trial)
+                            {
+                                Application.DoEvents();
+                                Application.OpenForms.OfType<Form>().ForEach(f =>
+                                {
+                                    try
+                                    {
+                                        f.Close();
+                                    }
+                                    catch (Exception)
+                                    {
+                                        // ignore
+                                    }
+                                });
+                            }
+                        }
+                    }
+                });
 
-            showDialog();
+                showForm();
 
-            // Join the asynchronous test operation so any exceptions are rethrown on this thread
-            test.Join();
+                // Join the asynchronous test operation so any exceptions are rethrown on this thread.
+                test.Join();
+            }
+            finally
+            {
+                form?.Close();
```

Comment:
📝 This call to `Close()` is unnecessary. It was already closed during the test above.

Reply:
In https://github.com/gitextensions/gitextensions/pull/7625#discussion_r372054764, you told me to add it.

Reply:
I have no concern to removed it.

Reply:
Ah, yes that is better. Main point is we only need one call to `Close()`.

Reply:
removed